### PR TITLE
test(sum): add more sum tests

### PR DESCRIPTION
**Pull request by giard-alexandre**
This should address the first 3 bullet points in https://github.com/reactivemarbles/DynamicData/discussions/1031#discussioncomment-15777487

### CODE REVIEW

File: src/DynamicData.Tests/AggregationTests/SumFixture.cs [hunk 1]
```diff
@@ -1,227 +1,786 @@
...
+            });
+
+            results.Error.Should().BeNull("no errors should have occurred");
+            results.HasCompleted.Should().BeFalse("the source can still publish notifications");
+            results.RecordedValues.Should().ContainSingle("a batched edit should produce exactly one sum emission")
+                .Which.Should().Be(60, "the sum of ages 10 + 20 + 30 is 60");
+        }
+
+        [Fact]
+        public void NoItemsAdded_NoSumEmitted()
```
Comment (JakenVeina) [inline src/DynamicData.Tests/AggregationTests/SumFixture.cs]: This reflects what the current behavior is? If so, it's one of the things that should be considered for a change. The most compelling argument, I think, is that `Enumerable.Sum()` emits `0` on an empty source, rather than, say, throwing an exception, so that's probably what consumers are going to expect, intuitively.

Not asking for a change, if this is the current behavior, that's what the test should reflect. Just musing about something to consider in the next phase.

Comment (JakenVeina) [inline src/DynamicData.Tests/AggregationTests/SumFixture.cs]: Me, I'd probably use `SourceIsEmpty` to describe the scenario here. It's slightly clearer about what's really being tested.

File: src/DynamicData.Tests/AggregationTests/SumFixture.ForCache.cs [hunk 2]
```diff
@@ -1,227 +1,786 @@
...
+                .ValidateSynchronization()
+                .RecordValues(out var results);
+
+            results.Error.Should().BeNull("no errors should have occurred");
+            results.HasCompleted.Should().BeFalse("the source can still publish notifications");
+            results.RecordedValues.Should().BeEmpty("no items were added so no sum values should have been emitted");
+        }
+
+        [Fact]
+        public void AllItemsRemoved_SumReturnsToZero()
```
Comment (JakenVeina) [inline src/DynamicData.Tests/AggregationTests/SumFixture.ForCache.cs:151]: Yeah, this definitely confirms that we should be emitting `0` upon subscription to an empty source.

File: src/DynamicData.Tests/AggregationTests/SumFixture.cs [hunk 3]
```diff
@@ -1,227 +1,786 @@
...
+            results.HasCompleted.Should().BeFalse("the source can still publish notifications");
+            results.RecordedValues.Should().ContainSingle("an AddRange produces a single changeset")
+                .Which.Should().Be(expectedSum, $"the sum of the first {itemCount} items should be {expectedSum}");
+        }
+
+        [Theory]
+        [InlineData(0, 50)]
+        [InlineData(1, 40)]
+        [InlineData(2, 30)]
+        public void ItemIsRemoved_SumDecreases(int removalIndex, int expectedSum)
```
Comment (JakenVeina) [inline src/DynamicData.Tests/AggregationTests/SumFixture.cs]: Probably `SumReflectsRemoval` works better, and matches what you've used elsewhere. This actually highlights that you should probably add test cases for negative values, for completeness' sake.

File: src/DynamicData.Tests/AggregationTests/SumFixture.cs [hunk 4]
```diff
@@ -1,227 +1,786 @@
...
+                .Sum(p => (float?)p.Age)
+                .RecordValues(out var results);
+
+            results.RecordedValues[^1].Should().Be(60F, "the nullable float sum of ages 10 + 20 + 30 is 60");
+        }
     }
 
-    [Fact]
-    public void AddedItemsContributeToSumDouble()
+    public class ListSource
```
Comment (JakenVeina) [inline src/DynamicData.Tests/AggregationTests/SumFixture.cs]: Definitely split these into separate files. To match other operators where we've got multiple overloads, I'd do `SumFixture.ForCache.cs` and `SumFixture.ForList.cs`. Check out (if I recall correctly) `ToObservableChangeSetFixture`.

File: src/DynamicData.Tests/AggregationTests/SumFixture.ForCache.cs [hunk 5]
```diff
@@ -1,227 +1,786 @@
...
+            source.Edit(updater => updater.Clear());
+
+            results.Error.Should().BeNull("no errors should have occurred");
+            results.HasCompleted.Should().BeFalse("the source can still publish notifications");
+            results.RecordedValues.Should().HaveCount(2, "one additional sum value should have been emitted after clearing");
+            results.RecordedValues[^1].Should().Be(0, "all items were removed so the sum should return to zero");
+        }
+
+        [Fact]
+        public void SourceCompletesAfterEmitting_CompletionPropagates()
```
Comment (JakenVeina) [inline src/DynamicData.Tests/AggregationTests/SumFixture.ForCache.cs:178]: Add equivalent tests for `SourceCompletesImmediately` and `SourceFailsImmediately`.

Quite often, these scenarios need a little bit of custom logic, to ensure correct behavior, especially for operators that emit an initial value, or that should always emit a value when the source emits one.

In the case of `Sum()`, a source can exist that emits an initial changeset, and then a completion, because no new item changes are allowed, and these would both occur immediately upon subscription. We'd want to make sure that `.Sum()` also emits an initial sum value, for those items, and then a completion, both immediately upon subscription.

File: src/DynamicData.Tests/AggregationTests/SumFixture.cs [hunk 6]
```diff
@@ -1,227 +1,786 @@
...
+                .RecordValues(out var results);
+
+            results.Error.Should().BeNull("no errors should have occurred");
+            results.HasCompleted.Should().BeFalse("the source can still publish notifications");
+            results.RecordedValues.Should().ContainSingle("one changeset was published containing all pre-existing items")
+                .Which.Should().Be(40, "null values should be treated as zero, so the sum should be 10 + 0 + 30 = 40");
+        }
+
+        [Fact]
+        public void ItemsAreAdded_SumIsCorrect_ForInt()
```
Comment (JakenVeina) [inline src/DynamicData.Tests/AggregationTests/SumFixture.cs]: Since we're targeting specific numeric types, we should probably add some test cases for `.MaxValue` and `.MinValue` to guarantee no overflows occurring where there shouldn't be. Maybe as new test cases to these existing tests, maybe as dedicated test methods.

**Review by JakenVeina — CHANGES_REQUESTED** [review]
Looks solid, overall.  A few nitpicks, and a few suggested new tests/test cases.

File: src/DynamicData.Tests/AggregationTests/SumFixture.cs (hunk 1 above)
Comment (giard-alexandre) [reply src/DynamicData.Tests/AggregationTests/SumFixture.cs]: I 100% agree with this and have hit up against this before. Sure, `StartWith(0)` is easy enough to add but yeah, it's confusing that Count doesnt default to 0.

**Comment by giard-alexandre** [comment]
Alright sorry for the extended delay in getting this pushed out, long story but I moved countries so free time was at a premium.

 @JakenVeina Everything mentioned above should have been addressed, I'll let you resolve the conversations if it's done to your liking!

**Review by JakenVeina — APPROVED** [review]
